File: SCRIPTS/ted_ali.pl
#!/usr/bin/env perl
use strict;
use warnings;
use LWP::UserAgent;
use JSON;
use Getopt::Long;

# Default to first positional argument for alignment file
my $alignment_file = $ARGV[0];
my $output_file = "TED";
my $mean_file = "TED_NUM";
my $verbose = 0;
my $force = 0;

GetOptions(
    "verbose" => \$verbose,
    "force"   => \$force,  # Add force option to overwrite existing files
    "help"    => \&usage
) or usage();

usage() unless $alignment_file;

# Check if output files already exist
if ((-e $output_file || -e $mean_file) && !$force) {
    print STDERR "Error: File '$output_file' or '$mean_file' already exists in this directory.\n";
    print STDERR "This indicates the script has already been run in this directory.\n";
    print STDERR "Use --force to overwrite the existing files if needed.\n";
    exit 1;
}

# Function to display usage
sub usage {
    print "Usage: $0 <alignment_file> [options]\n";
    print "Generates a report of TED domains for proteins in a Pfam alignment.\n\n";
    print "Output format (similar to Pfam arch file):\n";
    print "    P12345.1/10-200  3 domains\n";
    print "        TED01       10      150     85.0\n";
    print "        TED02      160      250     45.0\n";
    print "\n";
    print "Options:\n";
    print "  --verbose    Show detailed progress information\n";
    print "  --force      Overwrite existing files if they exist\n";
    print "  --help       Display this help message\n";
    exit 1;
}

# Function to strip version from UniProt accession
sub strip_version {
    my $acc = shift;
    # Remove version number if present (e.g., O34341.2 -> O34341)
    $acc =~ s/\.\d+$//;
    return $acc;
}

# Parse alignment file to extract sequence IDs
print STDERR "Parsing alignment file...\n" if $verbose;
my %sequences;
open(my $fh, '<', $alignment_file) or die "Cannot open $alignment_file: $!";
while (<$fh>) {
    chomp;
    if (/^(\S+)\/(\d+)-(\d+)/) {
        my $id = $1;
        my $uniprot_acc = strip_version($id);
        $sequences{$id} = {
            uniprot => $id,
            uniprot_no_version => $uniprot_acc,
            pfam_start => $2,
            pfam_end => $3,
            ted_domains => []
        };
    }
}
close($fh);

my $seq_count = scalar(keys %sequences);
print STDERR "Found $seq_count sequences in alignment.\n" if $verbose;

# Initialize HTTP client
my $ua = LWP::UserAgent->new;
$ua->timeout(30);
$ua->agent('Mozilla/5.0');

# Process each sequence and collect results
my $processed = 0;
my $with_ted = 0;
my $total_domains = 0;
my $errors = 0;

# Store results grouped by protein
my %protein_results;

foreach my $seq_id (sort keys %sequences) {
    $processed++;
    my $query_acc = $sequences{$seq_id}{uniprot_no_version};

    # Only print detailed progress in verbose mode
    if ($verbose) {
        print STDERR "Fetching TED domains for $seq_id (using $query_acc for TED query)...\n";
        print STDERR "Progress: $processed/$seq_count sequences processed\n";
    }

    my $ted_api_url = "https://ted.cathdb.info/api/v1/uniprot/summary/$query_acc?skip=0&limit=100";
    my $ted_response = $ua->get($ted_api_url);

    if (!$ted_response->is_success) {
        warn "Warning: Failed to fetch domain information from TED for $query_acc.\n" if $verbose;
        $errors++;
        next;
    }

    my $domain_data;
    eval {
        $domain_data = decode_json($ted_response->content);
    };

    if ($@ || !$domain_data) {
        warn "Warning: Failed to parse domain information for $query_acc.\n" if $verbose;
        $errors++;
        next;
    }

    my $domain_count = $domain_data->{count} || 0;
    if ($domain_count == 0) {
        # Skip proteins with no TED domains
        next;
    }

    $with_ted++;
    $total_domains += $domain_count;

    # Collect domains for this protein
    my @domain_list;

    foreach my $domain (@{$domain_data->{data}}) {
        my $ted_id = $domain->{ted_id};
        my $chopping = $domain->{chopping};

        # Extract just the TED suffix (e.g., TED01 from AF-xxx-TED01)
        my $ted_suffix = $ted_id;
        if ($ted_id =~ /(TED\d+)$/) {
            $ted_suffix = $1;
        }

        # Calculate overlap with Pfam domain
        my $pfam_start = $sequences{$seq_id}{pfam_start};
        my $pfam_end = $sequences{$seq_id}{pfam_end};

        # Parse domain boundaries
        my @segments = split(/_/, $chopping);
        my ($ted_start, $ted_end);

        if (@segments == 1) {
            # Single segment domain
            ($ted_start, $ted_end) = ($chopping =~ /^(\d+)-(\d+)$/);
        } else {
            # Multi-segment domain
            my $first_segment = $segments[0];
            my $last_segment = $segments[-1];
            ($ted_start) = ($first_segment =~ /^(\d+)/);
            ($ted_end) = ($last_segment =~ /(\d+)$/);
        }

        $ted_start = int($ted_start);
        $ted_end = int($ted_end);

        # Calculate overlap percentage with Pfam domain
        my $overlap_percent = 0;
        if ($ted_start <= $pfam_end && $ted_end >= $pfam_start) {
            my $pfam_length = $pfam_end - $pfam_start + 1;
            my $overlap_start = $pfam_start > $ted_start ? $pfam_start : $ted_start;
            my $overlap_end = $pfam_end < $ted_end ? $pfam_end : $ted_end;
            my $overlap_length = $overlap_end - $overlap_start + 1;
            $overlap_percent = ($overlap_length / $pfam_length) * 100;
        }

        push @domain_list, {
            ted_id => $ted_suffix,
            start => $ted_start,
            end => $ted_end,
            overlap => $overlap_percent
        };
    }

    $protein_results{$seq_id} = {
        domain_count => $domain_count,
        domains => \@domain_list
    };
}

# Write output file in arch-like format
open(my $out_fh, '>', $output_file) or die "Cannot open output file $output_file for writing: $!";

foreach my $seq_id (sort keys %protein_results) {
    my $result = $protein_results{$seq_id};
    my $domain_count = $result->{domain_count};

    # Summary line for the protein
    my $domain_word = $domain_count == 1 ? "domain" : "domains";
    print $out_fh "$seq_id\t$domain_count $domain_word\n";

    # One line per domain, indented
    foreach my $d (@{$result->{domains}}) {
        printf $out_fh "\t%s\t%d\t%d\t%.1f\n",
            $d->{ted_id}, $d->{start}, $d->{end}, $d->{overlap};
    }
}

close($out_fh);

# Calculate the mean domains per protein
my $mean_domains = $with_ted > 0 ? $total_domains / $with_ted : 0;

# Write the mean domains per protein to TED_NUM file
open(my $num_fh, '>', $mean_file) or die "Cannot open file $mean_file for writing: $!";
printf $num_fh "%.2f\n", $mean_domains;
close($num_fh);

# Print summary to STDERR
print STDERR "\n=== TED DOMAIN REPORT SUMMARY ===\n";
print STDERR "Total sequences analyzed: $seq_count\n";
print STDERR "Sequences with TED domains: $with_ted\n";
print STDERR "Total TED domains found: $total_domains\n";
print STDERR "Mean domains per protein: ", sprintf("%.2f", $mean_domains), "\n";
print STDERR "API or parsing errors: $errors\n";
print STDERR "Results written to file: $output_file\n";
print STDERR "Mean domains per protein written to: $mean_file\n";

File: SCRIPTS/ted_ali.py
#!/usr/bin/env python3
"""
Generate a report of TED domains for proteins in a Pfam alignment.

This script replaces ted_ali.pl with a Python implementation that uses
local TED data instead of API calls.

Usage:
    ted_ali.py <alignment_file> [options]

    ted_ali.py SEED
    ted_ali.py SEED --verbose
    ted_ali.py SEED --force

Output files:
    TED      - Tab-separated report of TED domains with overlap information
    TED_NUM  - Mean TED domains per protein (single number)

The TED file contains one line per TED domain found, with columns:
    Protein_Accession  TED_Domains  Domain_Count  Domain_IDs  Domain_Ranges  Overlap_with_Pfam
"""

import argparse
import os
import re
import sys
from typing import Dict, List, Tuple, Optional

# Import our local TED module
from ted_local import TEDLocal


def parse_arguments():
    """Parse command line arguments."""
    parser = argparse.ArgumentParser(
        description='Generate a report of TED domains for proteins in a Pfam alignment.'
    )
    parser.add_argument(
        'alignment_file',
        help='Path to the alignment file (e.g., SEED)'
    )
    parser.add_argument(
        '--verbose', '-v',
        action='store_true',
        help='Show detailed progress information'
    )
    parser.add_argument(
        '--force', '-f',
        action='store_true',
        help='Overwrite existing output files'
    )
    parser.add_argument(
        '--output', '-o',
        default='TED',
        help='Output file name (default: TED)'
    )
    parser.add_argument(
        '--mean-file', '-m',
        default='TED_NUM',
        help='Mean domains output file name (default: TED_NUM)'
    )
    parser.add_argument(
        '--ted-data',
        default=None,
        help='Path to local TED data file (uses default if not specified)'
    )

    return parser.parse_args()


def strip_version(accession: str) -> str:
    """
    Strip version number from UniProt accession.

    Args:
        accession: UniProt accession possibly with version (e.g., 'O34341.2')

    Returns:
        Accession without version (e.g., 'O34341')
    """
    return accession.split('.')[0]


def parse_alignment_file(alignment_file: str, verbose: bool = False) -> Dict[str, Dict]:
    """
    Parse alignment file to extract sequence IDs and regions.

    Args:
        alignment_file: Path to alignment file
        verbose: Print progress information

    Returns:
        Dict mapping sequence IDs to their info (uniprot, pfam_start, pfam_end)
    """
    if verbose:
        print("Parsing alignment file...", file=sys.stderr)

    sequences = {}

    with open(alignment_file, 'r') as f:
        for line in f:
            line = line.strip()
            if not line:
                continue

            # Match lines like: O34341.2/15-100 or P12345/1-200
            match = re.match(r'^(\S+)/(\d+)-(\d+)', line)
            if match:
                seq_id = match.group(0).split()[0]  # Get the full ID with region
                uniprot_with_version = match.group(1)
                pfam_start = int(match.group(2))
                pfam_end = int(match.group(3))

                sequences[seq_id] = {
                    'uniprot': uniprot_with_version,
                    'uniprot_no_version': strip_version(uniprot_with_version),
                    'pfam_start': pfam_start,
                    'pfam_end': pfam_end
                }

    if verbose:
        print(f"Found {len(sequences)} sequences in alignment.", file=sys.stderr)

    return sequences


def calculate_overlap_percentage(
    ted_start: int,
    ted_end: int,
    pfam_start: int,
    pfam_end: int
) -> float:
    """
    Calculate the percentage of the Pfam region covered by the TED domain.

    Args:
        ted_start: TED domain start position
        ted_end: TED domain end position
        pfam_start: Pfam region start position
        pfam_end: Pfam region end position

    Returns:
        Overlap percentage (0-100) relative to Pfam region length
    """
    # Check if there's any overlap
    if ted_start > pfam_end or ted_end < pfam_start:
        return 0.0

    # Calculate overlap
    overlap_start = max(ted_start, pfam_start)
    overlap_end = min(ted_end, pfam_end)
    overlap_length = overlap_end - overlap_start + 1

    # Calculate percentage relative to Pfam region
    pfam_length = pfam_end - pfam_start + 1
    overlap_percent = (overlap_length / pfam_length) * 100

    return overlap_percent


def main():
    args = parse_arguments()

    # Check input file exists
    if not os.path.exists(args.alignment_file):
        print(f"Error: Alignment file not found: {args.alignment_file}", file=sys.stderr)
        sys.exit(1)

    # Check if output files already exist
    if (os.path.exists(args.output) or os.path.exists(args.mean_file)) and not args.force:
        print(f"Error: Output file '{args.output}' or '{args.mean_file}' already exists.",
              file=sys.stderr)
        print("This indicates the script has already been run in this directory.",
              file=sys.stderr)
        print("Use --force to overwrite the existing files if needed.", file=sys.stderr)
        sys.exit(1)

    # Parse alignment file
    sequences = parse_alignment_file(args.alignment_file, args.verbose)

    if not sequences:
        print("Error: No sequences found in alignment file.", file=sys.stderr)
        sys.exit(1)

    # Initialize TED local data access
    if args.verbose:
        print("Loading TED domain data...", file=sys.stderr)

    try:
        ted = TEDLocal(args.ted_data)
    except FileNotFoundError as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)

    # Process sequences and collect results
    seq_count = len(sequences)
    processed = 0
    with_ted = 0
    total_domains = 0
    errors = 0

    results = []

    for seq_id in sorted(sequences.keys()):
        processed += 1
        seq_info = sequences[seq_id]
        query_acc = seq_info['uniprot_no_version']

        if args.verbose:
            print(f"Fetching TED domains for {seq_id} (using {query_acc})...",
                  file=sys.stderr)
            print(f"Progress: {processed}/{seq_count} sequences processed",
                  file=sys.stderr)

        # Get TED domains from local data
        try:
            domains = ted.get_domains(query_acc)
        except Exception as e:
            if args.verbose:
                print(f"Warning: Error getting domains for {query_acc}: {e}",
                      file=sys.stderr)
            errors += 1
            continue

        domain_count = len(domains)
        if domain_count == 0:
            # Skip proteins with no TED domains
            continue

        with_ted += 1
        total_domains += domain_count

        # Process each domain separately
        pfam_start = seq_info['pfam_start']
        pfam_end = seq_info['pfam_end']

        for domain in domains:
            ted_id = domain['ted_id']
            ted_suffix = domain['ted_suffix']
            chopping = domain['chopping']
            segments = domain['segments']

            # Calculate overall domain boundaries
            if segments:
                ted_start = min(seg['start'] for seg in segments)
                ted_end = max(seg['end'] for seg in segments)
            else:
                continue

            # Calculate overlap percentage
            overlap_percent = calculate_overlap_percentage(
                ted_start, ted_end, pfam_start, pfam_end
            )

            # Format overlap as percentage string
            if overlap_percent > 0:
                overlap_str = f"{overlap_percent:.1f}%"
            else:
                overlap_str = "0"

            results.append({
                'seq_id': seq_id,
                'ted_info': f"{ted_suffix}:{chopping}",
                'domain_count': 1,  # Each line reports one domain
                'ted_id': ted_suffix,
                'domain_range': f"{ted_start}-{ted_end}",
                'overlap': overlap_str
            })

    # Write output file
    with open(args.output, 'w') as f:
        # Header
        f.write("Protein_Accession\tTED_Domains\tDomain_Count\tDomain_IDs\t"
                "Domain_Ranges\tOverlap_with_Pfam\n")

        # Data
        for r in results:
            f.write(f"{r['seq_id']}\t{r['ted_info']}\t{r['domain_count']}\t"
                    f"{r['ted_id']}\t{r['domain_range']}\t{r['overlap']}\n")

    # Calculate mean domains per protein
    mean_domains = total_domains / with_ted if with_ted > 0 else 0

    # Write mean file
    with open(args.mean_file, 'w') as f:
        f.write(f"{mean_domains:.2f}\n")

    # Print summary
    print(f"\n=== TED DOMAIN REPORT SUMMARY ===", file=sys.stderr)
    print(f"Total sequences analyzed: {seq_count}", file=sys.stderr)
    print(f"Sequences with TED domains: {with_ted}", file=sys.stderr)
    print(f"Total TED domains found: {total_domains}", file=sys.stderr)
    print(f"Mean domains per protein: {mean_domains:.2f}", file=sys.stderr)
    print(f"Errors: {errors}", file=sys.stderr)
    print(f"Results written to file: {args.output}", file=sys.stderr)
    print(f"Mean domains per protein written to: {args.mean_file}", file=sys.stderr)


if __name__ == '__main__':
    main()
